### DISCUSSION TITLE
Bring ActiveRecord::Normalization to ActiveModel

### DISCUSSION THREAD

**Discussion by tmaier** [rails/rails#53793] [Ideas]
Rails 7.1 introduced [`ActiveRecord::Normalization`](https://github.com/rails/rails/blob/v8.0.0/activerecord/lib/active_record/normalization.rb) - a powerful feature that elegantly handles input normalization for database-backed models. It provides a clean, declarative way to ensure data consistency right at the attribute level.

However, **Rails applications aren't built on ActiveRecord alone**. At their core, many Ruby objects in a Rails app inherit behavior from ActiveModel - the foundation that provides validations, callbacks, and other model-like functionality without database persistence.

These **ActiveModel objects** (like Form Objects, Value Objects, or API Resource objects) **often need the same kind of input normalization**, but currently can't benefit from this new feature. For example:

- Value Objects need consistent formatting (phone numbers, ISBNs)
- Form Objects require input sanitization
- API Resource objects need to normalize data from external sources

I **propose extracting the core normalization features from ActiveRecord into ActiveModel**. This would make this powerful feature available to all objects that include ActiveModel, while keeping ActiveRecord-specific enhancements where they belong.

Here's how it could look:

```ruby
class PhoneNumber
  include ActiveModel::Model
  include ActiveModel::Normalization  # 🎯 Proposed

  attr_accessor :number, :country_code
  
  normalizes :number, with: -> num { num.delete("^0-9") }
  normalizes :country_code, with: -> code { code.to_s.delete("^0-9+") }
end
```

This would require some refactoring to separate ActiveRecord-specific functionality from the core normalization logic, but would bring consistent data normalization to more parts of our Rails applications. What are your thoughts on this approach?

**Comment by kasalava** [comment]
❤️

**Comment by flavorjones** [comment]
Thank you for starting this thread! I went through and re-read the original PR that introduced normalization, https://github.com/rails/rails/pull/43945, and in particular found a [relevant comment](https://github.com/rails/rails/pull/43945#issuecomment-1000918527) from @jonathanhefner:

> > could the normalizes class method be defined at the Active Model level
>
> I would very much like to, but the [type decoration mechanism for Active Record attributes](https://github.com/rails/rails/blob/35a5f22e1b7802538230387395ac271c1690848b/activerecord/lib/active_record/attributes.rb#L214-L229) does not exist [for Active Model attributes](https://github.com/rails/rails/blob/35a5f22e1b7802538230387395ac271c1690848b/activemodel/lib/active_model/attributes.rb#L22-L23)

The good news is that the type decoration mechanism was moved to ActiveModel in https://github.com/rails/rails/pull/44666 (specifically e0a55b03) so that shouldn't be a blocker anymore.

@jonathanhefner would you have any objection to reviewing a PR like the one described?

**Reply by jonathanhefner** [reply]
Indeed, part of the intent of #44666 was to enable porting `ActiveRecord::Normalization` to `ActiveModel`. :smiley:  

Looking through `ActiveRecord::Normalization`, I think all necessary dependencies are present in `ActiveModel::AttributeRegistration`, so porting it should now be straightforward.  I would be happy to review such a PR.

**Reply by seanpdoyle** [reply]
I've opened https://github.com/rails/rails/pull/53887 to explore this. I've also opened https://github.com/rails/rails/pull/53886 as a somewhat related change, but Normalization can occur with `send(:"#{name}=", send(name))` until `#[]` and `#[]=` exist.

**Comment by seanpdoyle** [comment]
https://github.com/rails/rails/pull/53887 has been merged into `main` 🎉 !

**Reply by tmaier** [reply]
Wow! Thank you! I did not expect this gets such a traction and gets resolved so quickly.

**Comment by thomaswitt** [comment]
@tmaier Thanks for the work on adding normalizes. I was really looking forward to use that and tried wiring it into a non–Active Record project (we’re using [Aws::Record](https://github.com/aws/aws-record-ruby) models which is backed by ActiveModel).

I hit a NoMethodError on decorate_attributes: the new module assumes the AR attribute decorator stack is present, but ActiveModel::Model/Aws::Record classes don’t provide those hooks.

It would be great if the API could either fall back to a plain ActiveModel implementation or document that it’s AR-only (because the title is maybe a bit misleading); otherwise consumers like us (running without ActiveRecord) can’t adopt it and have to keep custom normalization concerns.

Happy to share more detail if it helps.

**Reply by skipkayhil** [reply]
Can you open an issue reproducing the problem? I'm not sure I understand how these tests can pass https://github.com/rails/rails/blob/513b6b1307484edee7904f3091cbc326ca7a9063/activemodel/test/cases/attributes/normalization_test.rb if something in `normalizes` depends on Active Record

**Reply by thomaswitt** [reply]
Thanks for the quick follow-up!

The ActiveModel test case you mentioned passes because the sample class mixes in `ActiveModel::Attributes`, which defines the decorator stack and, in particular, the `decorate_attributes` hook that normalizes calls. As said, we are using [aws-record-ruby](https://github.com/aws/aws-record-ruby) and don't include that module. Aws::Record defines its own string_attr, integer_attr, etc. (a common pattern) which map into DynamoDB's attributes as macros (they're basically marshallers with setter/getters but never touching the AttributeSet of ActiveModel,  so there’s no decorate_attributes method to hook into. To make it work, I guess we’d have to redefine every attribute twice (attribute :foo for ActiveModel, plus string_attr :foo for Aws::Record) or build a bridge layer that makes Aws::Record emit the decorators ActiveModel expects.

Here’s some example code:

```
require 'active_model'
require 'aws-record'

class NormalizedAwsRecord
  include ActiveModel::Model
  include Aws::Record
  include ActiveModel::Attributes::Normalization

  string_attr :name
  normalizes :name, with: ->(value) { value&.strip }
end
```

When you instantiate this class, that triggers:
>  NoMethodError: undefined method `decorate_attributes' for NormalizedAwsRecord:Class

So the macro works great when ActiveModel::Attributes is present (matching the tests), but it currently can’t be used with Aws::Record because that stack doesn’t expose the decorator helpers.

I understand that this is a "[hey, it's the problem of the other gem](https://github.com/aws/aws-record-ruby/issues/152)" situation, but I just wanted to make you aware of that. As far as I know, some other Gems are also built on a similar pattern. The interesting thing is just, that I include for example
```
  include ActiveModel::Model
  include ActiveModel::Validations
  include ActiveModel::Validations::Callbacks
```

and that just works like a charm, giving me methods like persisted?, new_record? and destroyed? as well as callbacks.
